File: crack-data-structures-and-algorithms/leetcode/permutations_q46.cpp
#include <algorithm>
#include <vector>

using namespace std;

// 核心思路
// 首先 sort 后保证元素按非递减序，这是我们第一个排列
// 然后不断调用 next_permu 来生成下一个排列
// next_permu 核心思路
// 1) 找到**最大**的一个 i 满足 arr[i] < arr[i+1]，注意这里 i 范围为 [0, lens-2]
// 如果找不到这样的 i 说明当前已经是最后一个排列！
// 2) 假设第一步能找到，这一步找到**最大**的一个 j 使得 arr[i] < arr[j]；这里 j 范围为 [0, len-1]
// 3) swap arr[i] <-> arr[j]
// 4) reverse arr[i+1...len-1] 即反转 i+1 开始到末尾的所有元素
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> result;
        do {
            result.push_back(nums);
        } while (next_permu(nums));

        return result;
    }

    bool next_permu(vector<int>& nums) {
        int l = -1;
        for (int i = 0; i < nums.size() - 1; ++i) {
            if (nums[i] < nums[i+1]) {
                l = i;
            }
        }

        if (l == -1) {
            return false;
        }

        int r = -1;
        for (int i = l + 1; i < nums.size(); ++i) {
            if (nums[l] < nums[i]) {
                r = i;
            }
        }

        swap(nums[l],nums[r]);
        reverse(next(nums.begin(), l + 1), nums.end());
        return true;
    }
};

File: crack-data-structures-and-algorithms/leetcode/python-impl/permutations_q46.py
# 核心思路
# 首先排序保证数组为递增序（题目保证每个元素都不同）
# 然后不断调用next-permutation生成序列
class Solution(object):
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        nums.sort()
        result = []
        while True:
            result.append(nums[:])
            if not next_perm(nums):
                break

        return result


def next_perm(nums):
    x, y = -1, -1
    for i in range(len(nums) - 1):
        if nums[i] < nums[i+1]:
            x = i

        if x != -1 and nums[x] < nums[i]:
            y = i

    # Have reached the last permutation.
    if x == -1:
        return False

    # Check the last element.
    if nums[-1] > nums[x]:
        y = len(nums) - 1

    nums[x], nums[y] = nums[y], nums[x]
    nums[x+1:] = reversed(nums[x+1:])

    return True